_source/_posts/2019-09-27-build-a-crud-app-with-jsf.md

```diff
@@ -0,0 +1,786 @@
+      xmlns:p="http://primefaces.org/ui">
+<h:head></h:head>
+<h:body>
+    <h:form>
+        <p:panel header="Update Book">
+            <p:panelGrid columns="1" layout="grid">
+                <p:outputLabel for="book-title" value="Enter new book title"/>
+                <p:inputText id="book-title" value='#{bookBean.bookTitle}'/>
+                <p:commandButton value="Update" action="#{bookBean.update}" ajax="false"/>
+            </p:panelGrid>
+            <input type="hidden" name="bookId" value='#{param.bookId}'/>
+            <!-- We will use this later
+            <input type="hidden" value="${_csrf.token}" name="${_csrf.parameterName}"/>
+            -->
+        </p:panel>
+    </h:form>
+</h:body>
+</html>
+```
+
+File: `src/main/webapp/result.xhtml` using a `p:dataList` instead of `ui:repeat`:
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- File: result.xhtml -->
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
+        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml"
+      xmlns:f="http://java.sun.com/jsf/core"
+      xmlns:h="http://java.sun.com/jsf/html"
+      xmlns:p="http://primefaces.org/ui">
+<h:head></h:head>
+<h:body>
+    <h:link outcome="book" value="Create Book"/>
+    <p:dataList value="#{bookList.booksAvailable}" var="book" type="ordered">
+        <f:facet name="header">
+            Book List
+        </f:facet>
+        #{book.bookTitle}
+        <h:outputText value=" ("/>
+        <p:link value="Edit" outcome="edit">
+            <f:param name="bookId" value="#{book.bookId}"/>
+        </p:link>
+        <h:outputText value=" | "/>
+        <p:link value="Delete" outcome="delete">
+            <f:param name="bookId" value="#{book.bookId}"/>
+        </p:link>
+        <h:outputText value=")"/>
+    </p:dataList>
+</h:body>
+</html>
+```
+
+## Run the Application with PrimeFaces Enabled
+
+Restart the app with `mvn package tomee:run`. The app will look a bit better now! Check out the book list:
+
+{% img blog/build-a-crud-app-with-jsf/prime-faces-list.png alt:"Book list with PrimeFaces" width:"600" %}{: .center-image }
+
+## Secure Your Application With Okta
+
+At the moment, anyone can access our awesome Book application and change the database.
+To prevent that, let's add a security layer to our application with the Spring Security library and authenticate users through Okta.
+
+First, [register for a forever-free developer account today!](https://developer.okta.com/signup/) When you're finished, complete the steps below to create an OIDC app.
+
+1. Log in to your developer account at [developer.okta.com](https://developer.okta.com/).
+2. Navigate to **Applications** and click on **Add Application**.
+3. Select **Web** and click **Next**.
+4. Give the application a name (.e.g., `Java JSF Secure CRUD`) 
+5. Add the following as Login redirect URI:
+    - `http://localhost:8080/login/oauth2/code/okta`
+5. Click **Done**
```

Comment:
I'd remove the periods from the first three bullets since the last three don't have them.